ada/extensions/src/libadalang-helpers.ads

```diff
@@ -54,21 +85,45 @@ package Libadalang.Helpers is
       Description : String;
       --  Description for the application. Will be used in the help string.
 
-      with procedure Process_Context_Before
-        (Ctx     : Analysis_Context;
-         Project : GNATCOLL.Projects.Project_Tree_Access) is null;
+      Enable_Parallelism : Boolean := False;
+      --  If True, add a -j/--jobs command line option to allow multiple jobs
+      --  to run in parallel. In this mode, we create one analysis context per
+      --  job, and the files to process are distributed on each job. All the
+      --  callbacks below are called concurrently in all jobs:
+      --
+      --  First, the main task calls App_Setup. Then all jobs start:
+      --
+      --    * Job 1 calls Job_Setup, then several Process_Units, then
+      --      Job_Tear_Down.
+      --
+      --    * Job 2 calls Job_Setup, then several Process_Units, then
+      --      Job_Tear_Down.
+      --
+      --    * Job 3 calls ...
+      --
+      --  Finally, once all jobs are done, the main task calls App_Tear_Down.
+
+      with procedure App_Setup
+        (Context : App_Context; Jobs : App_Job_Context_Array) is null;
+      --  This procedure is called right after command line options are parsed,
+      --  the project is loaded (if present) and the list of files to process
+      --  is computed.
+
+      with procedure Job_Setup (Context : App_Job_Context) is null;
       --  This procedure will be called right before going through units.  If a
       --  project was loaded, Project refers to it, otherwise it is null.
 
       with procedure Process_Unit
-        (Unit : Analysis_Unit) is null;
+        (Context : App_Job_Context; Unit : Analysis_Unit) is null;
```

Comment:
OK, so this obsoletes my previous comment :)

Reply:
:-)